flink-connector-aws/flink-connector-dynamodb/src/main/java/org/apache/flink/connector/dynamodb/source/reader/PollingDynamoDbStreamsShardSplitReader.java

```diff
@@ -0,0 +1,160 @@
+
+import software.amazon.awssdk.services.dynamodb.model.GetRecordsResponse;
+import software.amazon.awssdk.services.dynamodb.model.Record;
+
+import javax.annotation.Nullable;
+
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.Collections;
+import java.util.Deque;
+import java.util.Iterator;
+import java.util.Set;
+
+import static java.util.Collections.singleton;
+
+/**
+ * An implementation of the SplitReader that periodically polls the DynamoDb stream to retrieve
+ * records.
+ */
+@Internal
+public class PollingDynamoDbStreamsShardSplitReader
+        implements SplitReader<Record, DynamoDbStreamsShardSplit> {
+
+    private static final RecordsWithSplitIds<Record> INCOMPLETE_SHARD_EMPTY_RECORDS =
+            new DynamoDbStreamRecordsWithSplitIds(Collections.emptyIterator(), null, false);
+
+    private final StreamProxy dynamodbStreams;
+
+    private final Deque<DynamoDbStreamsShardSplitState> assignedSplits = new ArrayDeque<>();
+
+    public PollingDynamoDbStreamsShardSplitReader(StreamProxy dynamodbStreamsProxy) {
+        this.dynamodbStreams = dynamodbStreamsProxy;
+    }
+
+    @Override
+    public RecordsWithSplitIds<Record> fetch() throws IOException {
+        DynamoDbStreamsShardSplitState splitState = assignedSplits.poll();
+        if (splitState == null) {
+            return INCOMPLETE_SHARD_EMPTY_RECORDS;
+        }
+
+        GetRecordsResponse getRecordsResponse =
+                dynamodbStreams.getRecords(
+                        splitState.getStreamArn(),
+                        splitState.getShardId(),
+                        splitState.getNextStartingPosition());
+        boolean isComplete =
+                getRecordsResponse == null || getRecordsResponse.nextShardIterator() == null;
+
+        if (hasNoRecords(getRecordsResponse)) {
+            if (isComplete) {
+                return new DynamoDbStreamRecordsWithSplitIds(
+                        Collections.emptyIterator(), splitState.getSplitId(), true);
+            } else {
+                assignedSplits.add(splitState);
+                return INCOMPLETE_SHARD_EMPTY_RECORDS;
+            }
+        }
+
+        splitState.setNextStartingPosition(
+                StartingPosition.continueFromSequenceNumber(
+                        getRecordsResponse
+                                .records()
+                                .get(getRecordsResponse.records().size() - 1)
+                                .dynamodb()
+                                .sequenceNumber()));
+
+        assignedSplits.add(splitState);
+        return new DynamoDbStreamRecordsWithSplitIds(
+                getRecordsResponse.records().iterator(), splitState.getSplitId(), isComplete);
+    }
+
+    private boolean hasNoRecords(GetRecordsResponse getRecordsResponse) {
+        return getRecordsResponse == null
+                || !getRecordsResponse.hasRecords()
+                || getRecordsResponse.records().isEmpty();
+    }
+
```

Comment:
These null checks are not needed anymore

Reply:
removed